bee-network/src/conns/manager.rs

```diff
@@ -0,0 +1,245 @@
+
+        Ok(Self {
+            peers,
+            banned_peers,
+            banned_addrs,
+            peer_listener,
+            internal_event_sender,
+        })
+    }
+}
+
+#[async_trait]
+impl<N: Node> Worker<N> for ConnectionManager {
+    type Config = ConnectionManagerConfig;
+    type Error = Infallible;
+
+    fn dependencies() -> &'static [TypeId] {
+        vec![TypeId::of::<PeerManager>()].leak()
+    }
+
+    async fn start(node: &mut N, config: Self::Config) -> Result<Self, Self::Error> {
+        let ConnectionManagerConfig {
+            peers,
+            banned_peers,
+            banned_addrs,
+            peer_listener,
+            internal_event_sender,
+            ..
+        } = config;
+
+        node.spawn::<Self, _, _>(|shutdown| async move {
+            info!("Listener started.");
+
+            let mut incoming = ShutdownStream::new(shutdown, peer_listener);
+
+            while let Some(Ok(listener_event)) = incoming.next().await {
+                //
+                if let Some((upgrade, peer_address)) = listener_event.into_upgrade() {
+                    // TODO: try again to move this block into its own function (beware: lifetime issues ahead!!!)
+
+                    // Prevent accepting from banned addresses.
+                    let peer_address_str = peer_address.to_string();
+                    if banned_addrs.contains(&peer_address_str) {
+                        trace!("Ignoring peer. Cause: {} is banned.", peer_address_str);
+                        NUM_LISTENER_EVENT_PROCESSING_ERRORS.fetch_add(1, Ordering::Relaxed);
+                        continue;
+                    }
+
+                    let (peer_id, muxer) = match upgrade.await {
+                        Ok(u) => u,
+                        Err(_) => {
+                            trace!("Ignoring peer. Cause: Handshake failed.");
+                            NUM_LISTENER_EVENT_PROCESSING_ERRORS.fetch_add(1, Ordering::Relaxed);
+                            continue;
+                        }
+                    };
+
+                    // Prevent accepting duplicate connections.
+                    if let Ok(connected) = peers.is(&peer_id, |_, state| state.is_connected()).await {
+                        if connected {
+                            trace!("Already connected to {}", peer_id);
+                            NUM_LISTENER_EVENT_PROCESSING_ERRORS.fetch_add(1, Ordering::Relaxed);
+                            continue;
+                        }
+                    }
+
+                    // Prevent accepting banned peers.
+                    if banned_peers.contains(&peer_id) {
+                        trace!("Ignoring peer. Cause: {} is banned.", peer_id);
+                        NUM_LISTENER_EVENT_PROCESSING_ERRORS.fetch_add(1, Ordering::Relaxed);
+                        continue;
+                    }
+
+                    let peer_info = if let Ok(peer_info) = peers.get_info(&peer_id).await {
```

Comment:
yea, maybe. I tried that already, but the order is slightly different due to the fact, that in one case you have the actuall peer id of the peer, and in the other case you start with an address only (mulitaddr).

Reply:
Thinking about abstracting this I had the idea of defining a `Filter` trait, where each of those checks are implemented as a concrete filter. Those then can be chained together in different order depending on the usecase. This would be very flexible as new filters could easily be added or removed.